src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -125712,6 +125817,49 @@ void M()
                 Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "location").WithLocation(18, 13));
         }
 
+        [Fact]
+        [WorkItem(46673, "https://github.com/dotnet/roslyn/issues/46673")]
+        public void CompareExchange_NamedArguments()
+        {
+            var source = @"
+#pragma warning disable 0436
+using System.Threading;
+namespace System.Threading
+{
+    public static class Interlocked
+    {
+        public static object? CompareExchange(ref object? location, object? value, object? comparand) { return location; }
+    }
+}
+
+class C
+{
+    void M()
+    {
+        object location1 = """";
+        Interlocked.CompareExchange(ref location1, comparand: """", value: null); // 1
+        
+        object location2 = """";
+        Interlocked.CompareExchange(ref location2, comparand: null, value: """");
+
+        object location3 = """";
+        Interlocked.CompareExchange(comparand: """", value: null, location: ref location3); // 2
+        
+        object location4 = """";
+        Interlocked.CompareExchange(comparand: null, value: """", location: ref location4);
+    }
+}
+";
+            var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (17,41): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         Interlocked.CompareExchange(ref location1, comparand: "", value: null); // 1
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "location1").WithLocation(17, 41),
+                // (23,79): warning CS8600: Converting null literal or possible null value to non-nullable type.
+                //         Interlocked.CompareExchange(comparand: "", value: null, location: ref location3); // 2
+                Diagnostic(ErrorCode.WRN_ConvertingNullableToNonNullable, "location3").WithLocation(23, 79));
+        }
```

Comment:
>} [](start = 8, length = 1)

Consider testing with optional parameters:
```C#
#nullable enable
using System.Threading;
namespace System.Threading
{
    class Interlocked
    {
        public static T CompareExchange<T>(ref T location1, T value, T comparand = default) where T : class => value;
    }
}
class C
{
    static void F(object target, object value)
    {
        Interlocked.CompareExchange(value: value, location1: ref target);
    }
}
```

Reply:
This already throws an IndexOutOfRangeException:

```csharp
#nullable enable
using System.Threading;
namespace System.Threading
{
    class Interlocked
    {
        public static T CompareExchange<T>(ref T location1, T value, T comparand = default) where T : class => value;
    }
}
class C
{
    static void F(object target, object value)
    {
        Interlocked.CompareExchange(ref target, null);
    }
}
```

Given that CompareExchange doesn't have optional arguments I don't know if it's worth the added complexity to fix it.

Reply:
Perhaps we should just bail in such a case to avoid the exception, but we won't get accurate diagnostics?

Reply:
> Perhaps we should just bail in such a case to avoid the exception, but we won't get accurate diagnostics?

Bailing out from `LearnFromCompareExchangeMethod()` and avoiding an exception seems fine to me.

---
In reply to: [491795175](https://github.com/dotnet/roslyn/pull/47523#discussion_r491795175) [](ancestors = 491795175)